asf_heat_pump_affordability/notebooks/gshp_analysis.py

```diff
@@ -0,0 +1,447 @@
+#
+# There are quite a lot of EPC records missing number of habitable rooms, this is slightly higher as a proportion of all records in Wales and Scotland.
+#
+# As such, we'll look at habitable rooms, then attempt to do something similar with floor sizes.
+
+# %%
+if cost_year_min == 2021:
+    # there is a single non-domestic observation for houses/bungalows installed 2021+, so we'll drop it for convenience.
+    idx = mcs_epc_data.loc[
+        lambda df: df["PROPERTY_TYPE"].isin(["House", "Bungalow"])
+        & (df["installation_type"] == "Non-Domestic")
+    ].index
+    mcs_epc_data = mcs_epc_data.drop(index=idx)
+
+# %%
+mcs_epc_data.loc[lambda df: df["PROPERTY_TYPE"].isin(["House", "Bungalow"])].shape
+
+# %%
+# Country variable
+mcs_epc_data["country"] = (
+    mcs_epc_data["lsoa11"].str[0].map({"E": "England", "W": "Wales", "S": "Scotland"})
+)
+mcs_epc_data.loc[
+    mcs_epc_data["PROPERTY_TYPE"].isin(["House", "Bungalow"])
+    & (mcs_epc_data["NUMBER_HABITABLE_ROOMS"] > 0)
+]["country"].value_counts()
+
+# %%
+mcs_epc_data.loc[
+    mcs_epc_data["PROPERTY_TYPE"].isin(["House", "Bungalow"])
+    & (mcs_epc_data["NUMBER_HABITABLE_ROOMS"] > 0)
+]["country"].value_counts() / mcs_epc_data.loc[
+    mcs_epc_data["PROPERTY_TYPE"].isin(["House", "Bungalow"]), "country"
+].value_counts()
+
+
+# %%
+# Create property size categories based on habitable rooms.
+def get_property_size(x):
+    if isinstance(x, pandas._libs.missing.NAType):
+        return pandas.NA
+    elif x == 0:
+        return pandas.NA
+    elif (x >= 1) & (x <= 3):
+        return "small"
+    elif (x >= 4) & (x <= 6):
+        return "medium"
+    elif (x >= 7) & (x <= 9):
+        return "large"
+    else:
+        return "very large"
+
+
+# Habitable rooms needs a sensible classification
+mcs_epc_data["property_size"] = mcs_epc_data["NUMBER_HABITABLE_ROOMS"].apply(
+    lambda x: get_property_size(x)
+)
+
+# %%
+mcs_epc_data.loc[
+    mcs_epc_data["PROPERTY_TYPE"].isin(["House", "Bungalow"]), "property_size"
+].value_counts(dropna=False) / mcs_epc_data["PROPERTY_TYPE"].isin(
+    ["House", "Bungalow"]
+).sum()
+
+# %%
+# Looks like NA on property size is likely to be new builds (2007+).
+test = mcs_epc_data.loc[mcs_epc_data["PROPERTY_TYPE"].isin(["House", "Bungalow"])].copy(
+    deep=True
+)
+test["property_size"] = test["property_size"].fillna("NA")
+pandas.crosstab(test["property_size"], test["CONSTRUCTION_AGE_BAND"])
+
+# %%
+test[test["property_size"] == "NA"]["installation_type"].value_counts()
+
+# %%
+test["installation_type"].value_counts()
+
+# %%
+# These records are thus likely to represent retrofits or EPC subsequent to the new build issuance.
+f, ax = pyplot.subplots(figsize=(8, 6))
+
+colours = {
+    "small": "#e41a1c",
+    "medium": "#377eb8",
+    "large": "#4daf4a",
+    "very large": "#984ea3",
+}
+mode_50hdi = []
+
+for property_size in [
```

Comment:
I think it would be useful to add some documentation to what's happening within each step of this loop for future reference.